### PR TITLE
add redis implementation

### PR DESCRIPTION
**Pull request by minrk**
based on the situation described in #155, it appears that redis is still the _only_ kv store with a well supported Python client.

After the refactor of the KV Store base class, adding a new kv store backend is very easy. This is the first one implemented since the refactor, and nothing needed to change (other than handling when `cleanup` may be async). So that worked well!

When I first ran the benchmarks, most things were fine, but `delete` was extremely slow and `O(number of routes)`. 

Screenshot 2024-01-18 at 14 31 34

It was implemented as:

```python
async for sub_key in self.redis.scan_iter(key + "*"):
    to_delete.append(sub_key)
await self.redis.delete(*to_delete)
```

However, redis performance recommendations suggested a lua script to run essentially the exact same code on the redis host itself. This seems to dramatically improve performance, and bring it in line or better than the other KV stores (see the benchmark notebook):

Screenshot 2024-01-18 at 16 16 27

I don't know if the Python delete implementation could be improved much, but the lua is simple enough and very quick, I don't see much reason it would need to be modified.

The only place where performance wasn't great was the add_route measurement with concurrency=50. I don't see 50 simultaneous transactions on the routing table as a super likely situation, so I'm not too worried about it. And at least the performance wasn't outrageous from a JupyterHub perspective (up to a few seconds at the longest). Performance was good up to concurrency=20, and significantly better than others for `get_all`.

Given the relative ease and support level of using redis, I think redis should be the _default_ traefik backend, and the one we recommend with this package.

TODO:

- [x] implement redis
- [x] tests
- [x] benchmarks
- [x] update CI for redis tests
- [x] update docs

### CODE REVIEW

**Comment by manics** [comment]
Sounds fine to me. If we make Redis the default backend can we also make it a mandatory requirement so it works out-of-the-box? The dependencies of redis-py are pretty minimal
https://github.com/redis/redis-py/blob/v5.0.1/setup.py#L38-L42

**Comment by minrk** [comment]
Yeah, I think that's a sensible idea.

**Comment by minrk** [comment]
I added docs, and ultimately decided _not_ to add it to default dependencies. The reason being that the most common place this package is used in tljh, where redis is _not_ needed. So it's not quite true that it's the _default_ backend, it's the default _distributed_ backend. I don't think documenting `pip install jupyterhub-traefik-proxy redis` is a huge burden, though w could have `extras` for the backends (especially since etcd/consul clients appear to be a moving target).

File: docs/source/install.md [hunk 1]
```diff
@@ -61,7 +61,9 @@
 
 If you want to use a key-value store to mediate configuration
 (mainly for use in distributed deployments, such as containers),
-you can get etcd or consul via their respective release pages:
+you can get The key-value stores via their respective installation pages:
```
Comment (consideRatio) [inline docs/source/install.md:64]:
```suggestion
you can get the key-value stores via their respective installation pages:
```

**Review by consideRatio — COMMENTED** [review]
I'd love to see that redis client become an extra dependency,  because there can be version constraints we want to encode and we can if we establish a pattern of installing redis client as part of installing this packge

File: jupyterhub_traefik_proxy/redis.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+    redis_url = Unicode(
+        'redis://localhost:6379', config=True, help="""The URL for the redis endpoint"""
+    )
+    redis_username = Unicode(config=True, help="The redis username")
+    redis_password = Unicode(config=True, help="The redis password")
+
+    redis_client_kwargs = Dict(
+        config=True,
+        help="Additional keyword arguments to pass through to the `redis.Redis` constructor",
```
Comment (consideRatio) [inline jupyterhub_traefik_proxy/redis.py:24]:
```suggestion
        help="Additional keyword arguments to pass through to the `redis.asyncio.Redis` constructor",
```

File: jupyterhub_traefik_proxy/redis.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        return Redis(**kwargs)
+
+    async def _cleanup(self):
+        f = super()._cleanup()
+        if f is not None:
+            await f
+        await self.redis.close()
+
+    def _setup_traefik_static_config(self):
+        self.log.debug("Setting up the redis provider in the static config")
```
Comment (consideRatio) [inline jupyterhub_traefik_proxy/redis.py:63]:
```suggestion
        self.log.debug("Setting up the redis provider in the traefik static config")
```

Too many pieces of config, being explicit is helpful - like the function name is!

**Review by consideRatio — APPROVED** [review]
I'd like to see redis client installed as an extra dep to help us manage constraints long term, but besides that this looks great from what i can say.

File: docs/source/install.md [hunk 4]
```diff
@@ -70,6 +72,7 @@ you can get etcd or consul via their respective release pages:
 Or, more likely, select the appropriate container image.
 You will also need to install a Python client for the Key-Value store of your choice:
 
+- `redis`
 - `etcdpy`
 - `python-consul2`
```
[the remark this answers is not in the document]
Comment (minrk) [reply docs/source/install.md:78]: updated, thanks

File: jupyterhub_traefik_proxy/redis.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        if to_delete:
+            self.log.debug("Deleting redis keys %s", to_delete)
+            futures.append(self.redis.delete(*to_delete))
+
+        await asyncio.gather(*futures)
+
+    async def _kv_get_tree(self, prefix):
+        """Return all data under prefix as a dict
+
+        Should probably use `unflatten_dict_from_kv`
```
Comment (GeorgianaElena) [inline jupyterhub_traefik_proxy/redis.py:153]: I see that `unflatten_dict_from_kv` is already used here, so maybe it's ok for this comment to go?

**Comment by consideRatio** [comment]
Everything seems okay to go, nice work!

**Review by GeorgianaElena — APPROVED** [review]
Wow, the performance difference with using lua for deleting routes. Also, it's so nice to see how straightforward it is right now, after the refactoring to add support for a new kv store.

It's a bummer that etcd and consul python clients don't have better support, so having this redis implementation and recommending it as the default sounds like a very good decision. Thank you!

File: jupyterhub_traefik_proxy/redis.py (hunk 5 above)
Comment (consideRatio) [reply jupyterhub_traefik_proxy/redis.py:153]: I'll add a commit about this in #231!

**Pull request merged**